xcm/xcm-executor/src/traits/on_response.rs

```diff
@@ -94,3 +98,57 @@ impl VersionChangeNotifier for () {
 		false
 	}
 }
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum FinishedQuery<BlockNumber> {
+	Response { response: Response, at: BlockNumber },
+	VersionNotification { origin: MultiLocation, is_active: bool },
+}
+
+/// The possible state of an XCM query response.
+#[derive(Debug, PartialEq, Eq)]
+pub enum QueryResponseStatus<BlockNumber> {
+	/// The response has arrived, and includes the inner Response and the block number it arrived at.
+	Finished(FinishedQuery<BlockNumber>),
+	/// The response has not yet arrived, the XCM might still be executing or the response might be in transit.
+	Pending,
+	/// No response with the given `QueryId` was found, or the response was already queried and removed from local storage.
+	NotFound,
+	/// Got an unexpected XCM version.
+	UnexpectedVersion,
+}
+
+/// Provides methods to expect responses from XCMs and query their status.
+pub trait XcmQueryHandler {
+	type QueryId: FullCodec + MaxEncodedLen + TypeInfo + Clone + Eq + PartialEq + Debug + Copy;
+	type BlockNumber;
+	type Error;
```

Comment:
For `QueryId` and `Error` generic types we have type within XCM protocol, `QueryId` and  `XcmError` for error, I think we can use them.